### PR TITLE
META-3040 Filter via purpose/persona(AccessControl) in Indexsearch

### PR DESCRIPTION
**Pull request by nikhilbonte21**
## Change description

> Description here

## Type of change
- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: graphdb/api/src/main/java/org/apache/atlas/repository/graphdb/AtlasIndexQuery.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 import org.apache.tinkerpop.gremlin.process.traversal.Order;
 
 import java.util.Iterator;
+import java.util.Map;
```
Comment (mehtaanshul) [inline graphdb/api/src/main/java/org/apache/atlas/repository/graphdb/AtlasIndexQuery.java:26]: No other change in this file, maybe we can remove this import

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
 package org.apache.atlas.repository.graphdb.janus;
 
 import java.util.Iterator;
+import java.util.Map;
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java:21]: is this import required?

File: repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java [hunk 3]
```diff
@@ -1016,6 +1051,28 @@ public AtlasSearchResult directIndexSearch(SearchParams searchParams) throws Atl
         return ret;
     }
 
+    private String getIndexName(IndexSearchParams params) throws AtlasBaseException {
+        if (StringUtils.isEmpty(params.getPersona()) && StringUtils.isEmpty(params.getPurpose())) {
+            return VERTEX_INDEX_NAME;
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java:1056]: Use Constants.VERTEX_INDEX

File: repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java [hunk 4]
```diff
@@ -1016,6 +1051,28 @@ public AtlasSearchResult directIndexSearch(SearchParams searchParams) throws Atl
...
+        }
+
+        String qualifiedName = "";
+        if (StringUtils.isNotEmpty(params.getPersona())) {
+            qualifiedName = params.getPersona();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java:1061]: understand this

File: repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java [hunk 5]
```diff
@@ -1016,6 +1051,28 @@ public AtlasSearchResult directIndexSearch(SearchParams searchParams) throws Atl
...
+        } else {
+            qualifiedName = params.getPurpose();
+        }
+
+        String[] parts = qualifiedName.split("/");
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java:1066]: better error handling if the client sends wrong persona/purpose

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java [hunk 6]
```diff
@@ -1683,6 +1685,12 @@ private EntityMutationResponse deleteVertices(Collection<AtlasVertex> deletionCa
         MetricRecorder metric = RequestContext.get().startMetricRecord("filterCategoryVertices");
         for (AtlasVertex vertex : deletionCandidates) {
             String typeName = getTypeName(vertex);
+
+            PreProcessor preProcessor = entityGraphMapper.getPreProcessor(typeName);
+            if (preProcessor != null) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java:1690]: add ensureNonAccessControlEntityType check as well

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    }
+
+    public void createPersona(PersonaContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("createPersona");
+        LOG.info("Creating Persona");
+        AtlasEntityWithExtInfo entityWithExtInfo = context.getPersonaExtInfo();
+        context.setCreateNewPersona(true);
+
+        try {
+            validateUniquenessByName(graph, getName(entityWithExtInfo.getEntity()), PERSONA_ENTITY_TYPE);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:119]: Why do we need this? won't this be automatically handled by qualifiedName + typeName

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+
+            String tenantId = getTenantId(context.getPersona());
+            if (StringUtils.isEmpty(tenantId)) {
+                tenantId = "tenant";
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:123]: tenantId = "default"

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+            }
+            entityWithExtInfo.getEntity().setAttribute(QUALIFIED_NAME, String.format("%s/%s", tenantId, getUUID()));
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:125]: isn't qualifiedName a mandatory attribute as part of request payload?

File: repository/src/main/java/org/apache/atlas/ranger/AtlasRangerService.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    public RangerRole createRangerRole(PersonaContext context) throws AtlasBaseException {
+        RangerRole ret;
+
+        try {
+            AtlasEntity personaEntity = context.getPersona();
+
+            RangerRole rangerRole = new RangerRole();
+            rangerRole.setName(getQualifiedName(personaEntity));
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/ranger/AtlasRangerService.java:68]: discuss

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+
+    private void cleanRoleFromAllRangerPolicies(AtlasEntityWithExtInfo personaExtInfo) throws AtlasBaseException {
+        AtlasEntity persona = personaExtInfo.getEntity();
+
+        List<RangerPolicy> rangerPolicies = fetchRangerPoliciesByLabel(atlasRangerService,
+                null,
+                null,
+                getPersonaLabel(persona.getGuid()));
+
+        List<String> allPolicyGuids = getPolicies(personaExtInfo).stream()
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:207]: these are policy labels right and not policyGuids?

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java (hunk 2 above)
Comment (nikhilbonte21) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java:21]: not required

File: repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java (hunk 3 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java:1056]: using same, added a static import

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java (hunk 6 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java:1690]: no need for this

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java (hunk 7 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:119]: this is uniqueness by name & not qualifiedName, so we have to handle it with our logic

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/callable/CleanRoleWorker.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+                long policyLabelCount = rangerPolicy.getPolicyLabels().stream().filter(x -> x.startsWith(LABEL_PREFIX_PERSONA)).count();
+                if (policyLabelCount == 0) {
+                    rangerPolicy.getPolicyLabels().remove(LABEL_TYPE_PERSONA);
+                }
+
+                atlasRangerService.updateRangerPolicy(rangerPolicy);
+            }
+        } catch (AtlasBaseException e) {
+            LOG.error("Failed to clean Ranger role from Ranger policies: {}", e.getMessage());
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/callable/CleanRoleWorker.java:65]: api should throw an error in this case

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/callable/CreateRangerPolicyWorker.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                }
+
+                List<String> labels = rangerPolicy.getPolicyLabels();
+                labels.addAll(AtlasPersonaUtil.getLabelsForPersonaPolicy(context.getPersona().getGuid(), context.getPersonaPolicy().getGuid()));
+                rangerPolicy.setPolicyLabels(labels);
+
+                ret = atlasRangerService.updateRangerPolicy(rangerPolicy);
+            }
+        } catch (AtlasBaseException e) {
+            LOG.error("Failed to create Ranger policies: {}", e.getMessage());
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/accesscontrol/persona/callable/CreateRangerPolicyWorker.java:65]: api should throw an error

File: repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java (hunk 4 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/discovery/EntityDiscoveryService.java:1061]: explained over a call

File: repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java (hunk 11 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java:207]: need to rename fields

**Pull request closed without merging**